File: 2021/01/rust/sonar-sweep.rs
fn main() {
    let part_1 = include_str!("../input.txt")
        .split("\n")
        .filter_map(|s| s.parse::<usize>().ok())
        .collect::<Vec<usize>>()
        .windows(2)
        .fold(0, |acc, win| if win[0] < win[1] { acc + 1 } else { acc } );

    println!("Part 1: {:?}", part_1);

    let part_2 = include_str!("../input.txt")
        .split("\n")
        .filter_map(|s| s.parse::<usize>().ok())
        .collect::<Vec<usize>>()
        .windows(4)
        .fold(0, |acc, win| {
            if win[0..3].iter().sum::<usize>() < win[1..4].iter().sum::<usize>() {
                acc + 1
            } else {
                acc
            }
        });

    println!("Part 2: {:?}", part_2);
}

File: 2021/01/typescript/sonar-sweep.ts
import fs from 'fs';

function getItems(): number[] {
    return fs
        .readFileSync('../input.txt', 'utf8')
        .split('\n')
        .map((s: string) => parseInt(s));
}

function part1(): number {
    return getItems().reduce((acc, curr, idx, arr) => {
        if (curr > arr[idx - 1]) {
            return acc + 1;
        }

        return acc;
    }, 0);
}

function part2(): number {
    return getItems().reduce((acc, _, idx, arr) => {
        if (
            arr
                .slice(idx, idx + 3)
                .filter((i) => i)
                .reduce((a, i) => a + i, 0) <
            arr
                .slice(idx + 1, idx + 4)
                .filter((i) => i)
                .reduce((a, i) => a + i, 0)
        ) {
            return acc + 1;
        }

        return acc;
    }, 0);
}

console.log('Part 1:', part1());
console.log('Part 2:', part2());
